src/runtime_src/core/edge/user/aie/graph.cpp

```diff
@@ -306,33 +314,110 @@ update_rtp(const char* port, const char* buffer, size_t size)
+        XAieTile_DmWriteWord(update_tile, start_addr, ((const uint32_t *)buffer)[i]);
         start_addr += 4;
     }
     if (remain) {
         uint32_t rdata = 0;
-        memcpy(&rdata, &((const u32*)buffer)[i], remain);
+        memcpy(&rdata, &((const uint32_t *)buffer)[i], remain);
         XAieTile_DmWriteWord(update_tile, start_addr, rdata);
     }
 
     /* update selector */
     XAieTile_DmWriteWord(selector_tile, rtp->selector_addr, selector);
 
     if (rtp->require_lock) {
-        /* release lock */
+        /* release lock, need to release lock even graph is not running */
         XAieTile_LockRelease(selector_tile, rtp->selector_lock_id, 1, 0XFFFFFFFF);
         XAieTile_LockRelease(update_tile, lock_id, 1, 0XFFFFFFFF);
     }
 }
 
+void
+graph_type::
+read_rtp(const char* port, char* buffer, size_t size)
+{
+    auto rtp = std::find_if(rtps.begin(), rtps.end(),
+            [port](rtp_type it) { return it.name.compare(port) == 0; });
+
+    if (rtp == rtps.end())
+      throw xrt_core::error(-EINVAL, "Can't read graph '" + name + "': RTP port '" + port + "' not found");
+
+    if (rtp->is_plrtp)
+      throw xrt_core::error(-EINVAL, "Can't read graph '" + name + "': not aie rtp port");
+
+    if (rtp->is_input)
+      throw xrt_core::error(-EINVAL, "Can't read graph '" + name + "': RTP port '" + port + "' is input");
+
+    /* If RTP port is connected, only support async update */
+    if (rtp->is_connected)
+      throw xrt_core::error(-EINVAL, "Can't read graph '" + name + "': connected read is not supported");
+
+    auto selector_pos = aieArray->getTilePos(rtp->selector_col, rtp->selector_row);
+    auto selector_tile = &(aieArray->tileArray.at(selector_pos));
+
+    /* Don't acquire selector lock for async RTP while graph is not running */
+    bool need_lock = rtp->require_lock && (
+        rtp->is_async && state == graph_state::running ||
+        !rtp->is_async
+        );
+
+    if (need_lock)
+        XAieTile_LockAcquire(selector_tile, rtp->selector_lock_id, 1, 0xFFFFFFFF);
+
+    uint32_t selector = XAieTile_DmReadWord(selector_tile, rtp->selector_addr);
+
+    int update_pos;
+    uint16_t lock_id;
+    uint64_t start_addr;
+    if (selector == 1) {
+        /* update pong buffer */
+        update_pos = aieArray->getTilePos(rtp->pong_col, rtp->pong_row);
+        lock_id = rtp->pong_lock_id;
+        start_addr = rtp->pong_addr;
+    } else {
+        /* update ping buffer */
+        update_pos = aieArray->getTilePos(rtp->ping_col, rtp->ping_row);
+        lock_id = rtp->ping_lock_id;
+        start_addr = rtp->ping_addr;
+    }
+
+    XAieGbl_Tile *update_tile = &(aieArray->tileArray.at(update_pos));
+    if (need_lock) {
+        XAieTile_LockAcquire(update_tile, lock_id, 1, 0xFFFFFFFF);
+	/* sync RTP release lock for write, async RTP relase lock for read */
+        XAieTile_LockRelease(selector_tile, rtp->selector_lock_id, (rtp->is_async ? 1 : 0), 0XFFFFFFFF);
+    }
```

Comment:
This looks suspiciously identical to update_rtp?  Can some code be shared?

Reply:
I've been trying to provide one interface with a flag update/read RTP. But it seems not easy.
1) Because of the direction, we have to use different AIE driver interface DmReadWord/DmWriteWord
2) The selector lock acquire/release sequence and parameter for update and read are different. 
    a) update needs to pass in 0 while read needs to pass in 1 
    b) for read, selector lock needs to be released before accessing the tile memory
    c) for update, needs to release selector lock even we don't acquiring it when graph is not running. 
3) Parameter for update_tile lock are different